File: java/src/main/java/io/github/skunkwerkx/hyperuuid/NativePlatform.java
package io.github.skunkwerkx.hyperuuid;

/**
 * Maps the running JVM's OS/arch to the RID-style directory (matching the C# wrapper's
 * {@code runtimes/{RID}/native/} convention) and filename the native library was built for.
 *
 * <p>A {@code .jar} has no package-manager-level platform selection the way NuGet's RID
 * folders or Python wheel tags do — one jar has to work everywhere, so it bundles every
 * platform's build under {@code /native/{rid}/{filename}} and this picks the right one at
 * runtime instead.
 */
final class NativePlatform {
    record Target(String rid, String libraryFileName) {}

    // Resolved once on first access to this class, mirroring the Java binding's `by lazy`
    // — a plain static field initializer already only runs on class-init, which happens on
    // first use.
    private static final Target CURRENT = detect();

    static Target current() {
        return CURRENT;
    }

    /** Classpath resource path for this platform's bundled native library. */
    static String resourcePath() {
        return "/native/" + CURRENT.rid() + "/" + CURRENT.libraryFileName();
    }

    private static Target detect() {
        String osName = System.getProperty("os.name").toLowerCase();
        String osArch = System.getProperty("os.arch").toLowerCase();
        boolean isArm = osArch.contains("aarch64") || osArch.contains("arm");

        if (osName.contains("win")) {
            return new Target(isArm ? "win-arm64" : "win-x64", "hyperuuid.dll");
        }
        if (osName.contains("mac") || osName.contains("darwin")) {
            return new Target(isArm ? "osx-arm64" : "osx-x64", "libhyperuuid.dylib");
        }
        if (osName.contains("linux")) {
            return new Target(isArm ? "linux-arm64" : "linux-x64", "libhyperuuid.so");
        }
        throw new IllegalStateException(
                "Unsupported platform for hyperuuid: os.name=" + osName + " os.arch=" + osArch);
    }

    private NativePlatform() {}
}

File: kotlin/src/main/kotlin/io/github/buvinghausen/hyperuuid/NativePlatform.kt
package io.github.buvinghausen.hyperuuid

/**
 * Maps the running JVM's OS/arch to the RID-style directory (matching the C# wrapper's
 * `runtimes/{RID}/native/` convention) and filename the native library was built for.
 *
 * A `.jar` has no package-manager-level platform selection the way NuGet's RID folders or
 * Python wheel tags do — one jar has to work everywhere, so it bundles every platform's build
 * under `/native/{rid}/{filename}` and this picks the right one at runtime instead.
 */
internal object NativePlatform {
    data class Target(val rid: String, val libraryFileName: String)

    val current: Target by lazy { detect() }

    /** Classpath resource path for this platform's bundled native library. */
    val resourcePath: String get() = "/native/${current.rid}/${current.libraryFileName}"

    private fun detect(): Target {
        val osName = System.getProperty("os.name").lowercase()
        val osArch = System.getProperty("os.arch").lowercase()
        val isArm = osArch.contains("aarch64") || osArch.contains("arm")

        return when {
            osName.contains("win") -> Target(if (isArm) "win-arm64" else "win-x64", "hyperuuid.dll")
            osName.contains("mac") || osName.contains("darwin") ->
                Target(if (isArm) "osx-arm64" else "osx-x64", "libhyperuuid.dylib")
            osName.contains("linux") -> Target(if (isArm) "linux-arm64" else "linux-x64", "libhyperuuid.so")
            else -> error("Unsupported platform for hyperuuid: os.name=$osName os.arch=$osArch")
        }
    }
}
